src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -0,0 +1,213 @@
+    Copyright (c) 2024, the clio developers.
+
+    Permission to use, copy, modify, and distribute this software for any
+    purpose with or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL,  DIRECT,  INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include "util/newconfig/ConfigDefinition.hpp"
+
+#include "util/Assert.hpp"
+#include "util/newconfig/Array.hpp"
+#include "util/newconfig/ArrayView.hpp"
+#include "util/newconfig/ConfigValue.hpp"
+#include "util/newconfig/ObjectView.hpp"
+
+#include <cstddef>
+#include <initializer_list>
+#include <optional>
+#include <stdexcept>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <variant>
+
+namespace util::config {
+/**
+ * @brief Full Clio Configuration definition.
+ *
+ * Specifies which keys are valid in Clio Config and provides default values if user's do not specify one. Those
+ * without default values must be present in the user's config file.
+ */
+static ClioConfigDefinition ClioConfig = ClioConfigDefinition{
+    {{"database.type", ConfigValue{ConfigType::String}.defaultValue("cassandra")},
+     {"database.cassandra.contact_points", ConfigValue{ConfigType::String}.defaultValue("localhost")},
+     {"database.cassandra.port", ConfigValue{ConfigType::Integer}},
+     {"database.cassandra.keyspace", ConfigValue{ConfigType::String}.defaultValue("clio")},
+     {"database.cassandra.replication_factor", ConfigValue{ConfigType::Integer}.defaultValue(3)},
+     {"database.cassandra.table_prefix", ConfigValue{ConfigType::String}.defaultValue("table_prefix")},
+     {"database.cassandra.max_write_requests_outstanding", ConfigValue{ConfigType::Integer}.defaultValue(10'000)},
+     {"database.cassandra.max_read_requests_outstanding", ConfigValue{ConfigType::Integer}.defaultValue(100'000)},
+     {"database.cassandra.threads",
+      ConfigValue{ConfigType::Integer}.defaultValue(static_cast<int>(std::thread::hardware_concurrency()))},
+     {"database.cassandra.core_connections_per_host", ConfigValue{ConfigType::Integer}.defaultValue(1)},
+     {"database.cassandra.queue_size_io", ConfigValue{ConfigType::Integer}.optional()},
+     {"database.cassandra.write_batch_size", ConfigValue{ConfigType::Integer}.defaultValue(20)},
+     {"etl_source.[].ip", Array{ConfigValue{ConfigType::String}.optional()}},
```

Comment:
I am not sure the ip is optional?

Reply:
I am a bit confused about the [] definition . When we use a.[].b, is a an array or b?
For etl_source.[].ip, i guess etlsource is array.
but for dos_guard.[].whitelist, should the whitelist is array?

Reply:
In the case of a.[].b, both getArrayView(a) and getArrayView(a.[].b) is a correct ArrayView, but getArrayView(a) is restricted to certain methods. This is because consider : 
"etl_sources": [
    {
       "ip": "127.0.0.1",
       "ws_port": "6006",
       "grpc_port": "50051"
    },
 {
       "ip": "idk",
       "ws_port": "1234",
       "grpc_port": "1111"
    }
        ],

Within each entry of etl_sources, “there’s a {}, which represents an object. Therefore, once you getArray(a), you can get the object at a certain index,  for example objectAt(0) will return the object with "ip": "127.0.0.1", "ws_port": "6006", etc., and objectAt(1) will return  "ip": "idk” …etc... 

In the other case of getArray(etl_sources.[].ip), it is just an arrayView in configDefinition, and so the .begin() can iterate through all the ip values, and valueAt(0) will fetch "127.0.0.1", and ivalueAt(1) will fetch "idk” because the 2nd ip has value of "idk"
so "dos_guard.[].whitelist" is an array.
Let me know if you have any questions or that there can be improvements. 😀.

Reply:
> I am not sure the ip is optional?

![Screenshot 2024-07-16 at 4 39 51 PM](https://github.com/user-attachments/assets/d077c433-4884-455c-b04e-b3b6b0c8c781)

I'm seeing in our codebase that it is optional. Let me know if I'm wrong 👍